docs/RFD-JobWorkerService.md

```diff
@@ -0,0 +1,366 @@
+Calling `Stream` for not-existing or not-started job returns empty `OutputReader`.
+
+
+* **func (\*Job) Stop() error** - attempt to gracefully shutdown the process.
+`Stop` will do nothing if the process has already completed or been killed.
+
+### API 
+The API is nothing more than just a wrapper around Library, but main API responsibility provide mTLS Authorization and using Library
+functionality via [gRPC](https://grpc.io/)
+
+#### Following functions considered to be used for API Server.
+
+* `Start` New Job
+  * verify user credentials via mTLS Authorization
+  * create new instance of the `Job` and assign new UUID 
+  * store `job` in sync.Map using UUID as key
+  * start command execution
+  * return UUID to the client
+* `Status` of the Job
+  * return current status of the `Job`
+* `Stream` latest Job output
+  * lookup `Job` by provided UUID
+  * request instance of `OutputReader` using library
+  * send steam output to the client/cli
+* `Stop` Job
+  * lookup `Job` by provided UUID
+  * call `Stop` function if `Job` was found, otherwise return error
+  
+  > **Not any job and it output removed from then server**. So, depend on the load and size of total output it would come to OOM situation sooner or later and [panic](https://go.dev/blog/defer-panic-and-recover).
+  > Probably we need to consider some API or process to clean completed jobs based on size, date of creation, frequency of usage or etc.
+  
+
+#### Security
+
+The client and API communicate via mTLS using [TLS 1.3 ](https://datatracker.ietf.org/doc/html/rfc8446) as the minimum version. The following cipher suites are supported:
+
+* TLS_AES_256_GCM_SHA384
+* TLS_CHACHA20_POLY1305_SHA256
+* TLS_AES_128_GCM_SHA256
+
+> Note: only 3 cipher suites enabled openssl by default: https://wiki.openssl.org/index.php/TLS1.3, and following 2 
+> below also can be used, but need to explicitly enabled: 
+>* TLS_AES_128_CCM_8_SHA256
+>* TLS_AES_128_CCM_SHA256
+
+> Since there are no requirements for older versions of TLS to be supported, TLS v1.3 will be made a minimum.
+
+#### Authorization
+All verification should be done on API level and API treats the client's common name (CN) in the certificate as the user's identity.
+Client not able interact with a job not created by them in any way (Stream, query Status, Stop for another user's job is not allowed).
+
+API not cache any user authorization data and such data should be provided for every API call.
+
+#### Proposed Protobuf
+Following [protobuf](https://protobuf.dev/programming-guides/proto3/) proposed for the client/server:
+
+```protobuf
+  syntax="proto3";
+
+  option go_package = "<package_name>/proto";
+
+  package proto;
+
+  service JobWorker {
+    rpc Start(JobCreateRequest) returns (JobResponse) {}
+    rpc Status(JobRequest) returns (JobStatusResponse) {}
+    rpc Stream(JobRequest) returns (stream OutputResponse) {}
+    rpc Stop(JobRequest) returns (JobStatusResponse) {}
```

Comment:
How a ongoing job will be stopped ?

Reply:
@smallinsky yes, it is also good point.

Job stopping on following cases:

1. Client send request to the server to stop command execution and if job still running forcibly terminate job

    _**Suggestion:**_
    _In this case all looks like clear and killing process support by standard library:_
    ```go
    // kill process:
    if err := cmd.Process.Kill(); err != nil {
        log.Fatal("failed to kill process: ", err)
    }
    ```
    
    _or in case of forcibly killing process we need to do something like this, because we know PID_
    ```go
    jobProcess := os.Process{
	    PID := cmd.Process.Pid
    }
    
    if err := jobProcess.Signal(syscall.SIGTERM); err != nil {
	    log.Fatalf("error sending signal to child process: %v", err)
    }
    ``'



2. Server terminating and should loop through the all running jobs, send notification to terminate and if job still running forcibly terminate job
    
    *_*Suggestion:**_
    _In this case cli need to do same as for Stopping command (see above), but for the list of the jobs._
    _And, of course, server should handle `syscall.SIGTERM`_
    ```go 
    sigCh := make(chan os.Signal, 1    
    signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM)
    wg := sync.WaitGroup{}
    wg.Add(1)
    go func() {
        s := <-sigCh
        log.Printf("got signal %v, attempting graceful shutdown", s)
                     
                    <STOP ALL JOBS>
    
        cancel()
        grpc.GracefulStop()
        // grpc.Stop() // leads to error while receiving stream response: rpc error: code = Unavailable desc = transport is closing
         wg.Done()
    }()
    ```

5. Server crashed _**[need to research if this ever possible to handle]**_

Reply:
@smallinsky added following to RFD.
![Screenshot 2024-08-26 at 11 06 58 PM](https://github.com/user-attachments/assets/9ff6a621-2346-424e-aac7-29b8a3b1e48c)